File: PThreads/PeterSmoking/PeterSmoking.cpp
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>  // Para a função sleep()
#include <time.h>    // Para a função time()

// Ingredientes
#define TABACO 0
#define PAPEL 1
#define FOSFORO 2
#define RODADAS 5

bool mesa[3] = {false, false, false};  // Ingredientes na mesa
bool ingredientes_disponiveis = false; // Indica se Peter já colocou os ingredientes
int rodadas_restantes = RODADAS;       // Número de rodadas restantes
bool rodadas_acabaram = false;         // Flag para indicar quando as rodadas terminam

pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;  // Mutex para sincronização

const char* ingredientes_nomes[] = {"Tabaco", "Papel", "Fósforo"};

void* fumante(void* arg) {
    int id = *(int*)arg;
    int ingrediente = id - 1;  // Assumindo que os IDs são 1, 2, 3

    bool continuar = true;
    while (continuar) {
        bool encontrou_ingredientes = false;

        pthread_mutex_lock(&mutex);
        // Verifica se os dois ingredientes que o fumante precisa estão na mesa
        if (ingredientes_disponiveis &&
            mesa[(ingrediente + 1) % 3] && mesa[(ingrediente + 2) % 3]) {
            printf("Fumante %d (que tem %s) pegou %s e %s e está fumando.\n",
                   id,
                   ingredientes_nomes[ingrediente],
                   ingredientes_nomes[(ingrediente + 1) % 3],
                   ingredientes_nomes[(ingrediente + 2) % 3]);

            // Limpa a mesa
            mesa[(ingrediente + 1) % 3] = false;
            mesa[(ingrediente + 2) % 3] = false;
            ingredientes_disponiveis = false;  // Ingredientes usados
            encontrou_ingredientes = true;
        }
        pthread_mutex_unlock(&mutex);

        if (encontrou_ingredientes) {
            sleep(rand() % 3 + 1);  // Simula o tempo de fumar
            printf("Fumante %d terminou de fumar.\n", id);
        } else {
            printf("Fumante %d (que tem %s) tentou pegar, mas os ingredientes corretos não estavam na mesa.\n",
                   id,
                   ingredientes_nomes[ingrediente]);
        }

        sleep(1);  // Fumante espera um pouco antes de verificar novamente

        pthread_mutex_lock(&mutex);
        if (rodadas_acabaram) continuar = false;  // Sai do loop quando as rodadas terminarem
        pthread_mutex_unlock(&mutex);
    }
    return NULL;
}

void* peter(void* arg) {
    bool continuar = true;
    while (continuar) {
        pthread_mutex_lock(&mutex);
        if (rodadas_restantes <= 0) {
            rodadas_acabaram = true;  // Marca que as rodadas acabaram
            continuar = false;  // Sai do loop
        }

        if (!ingredientes_disponiveis && rodadas_restantes > 0) {
            // Peter escolhe dois ingredientes aleatórios
            int ingrediente1 = rand() % 3;
            int ingrediente2 = (ingrediente1 + 1 + rand() % 2) % 3;

            mesa[ingrediente1] = true;
            mesa[ingrediente2] = true;
            ingredientes_disponiveis = true;

            printf("Peter colocou %s e %s na mesa. (Rodada %d)\n",
                   ingredientes_nomes[ingrediente1],
                   ingredientes_nomes[ingrediente2],
                   RODADAS - rodadas_restantes + 1);

            rodadas_restantes--;  // Diminui o número de rodadas restantes
        }
        pthread_mutex_unlock(&mutex);

        sleep(1);  // Peter coloca ingredientes periodicamente
    }

    printf("Peter terminou de colocar ingredientes após 5 rodadas.\n");
    return NULL;
}

int main() {
    srand(time(NULL));  // Inicializa o gerador de números aleatórios

    pthread_t thread_peter;
    pthread_t fumantes[3];
    int ids[3] = {1, 2, 3};

    // Cria a thread de Peter
    pthread_create(&thread_peter, NULL, peter, NULL);

    // Cria as threads dos fumantes
    for (int i = 0; i < 3; i++) {
        pthread_create(&fumantes[i], NULL, fumante, &ids[i]);
    }

    // Espera a thread de Peter terminar
    pthread_join(thread_peter, NULL);

    // Espera as threads dos fumantes terminarem
    for (int i = 0; i < 3; i++) {
        pthread_join(fumantes[i], NULL);
    }

    printf("Simulação concluída após 5 rodadas.\n");

    return 0;
}

File: Golang/PeterSmoking/PeterSmoking.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

type Ingredient int

const (
	Tabaco Ingredient = iota
	Papel
	Fosforo
)

var ingredientNames = []string{"Tabaco", "Papel", "Fósforo"}

const ROUNDS = 5

type Mesa struct {
	mutex            sync.Mutex
	cond             *sync.Cond
	ingredientes     [3]bool
	disponivel       bool
	rodadasRestantes int
	rodadasAcabaram  bool
}

func NewMesa() *Mesa {
	m := &Mesa{rodadasRestantes: ROUNDS}
	m.cond = sync.NewCond(&m.mutex)
	return m
}

func (m *Mesa) colocarIngredientes() {
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < ROUNDS; i++ {
		m.mutex.Lock()
		for m.disponivel {
			m.cond.Wait()
		}

		// Escolhe dois ingredientes aleatórios
		ing1 := rand.Intn(3)
		ing2 := (ing1 + 1 + rand.Intn(2)) % 3

		m.ingredientes[ing1] = true
		m.ingredientes[ing2] = true
		m.disponivel = true

		fmt.Printf("Peter colocou %s e %s na mesa. (Rodada %d)\n",
			ingredientNames[ing1], ingredientNames[ing2], i+1)

		m.cond.Broadcast()
		m.mutex.Unlock()
		time.Sleep(time.Second)
	}

	// Marca que o jogo acabou e acorda todos os fumantes
	m.mutex.Lock()
	m.rodadasAcabaram = true
	m.cond.Broadcast()
	m.mutex.Unlock()
}

func (m *Mesa) fumante(id int, meuIngrediente Ingredient, wg *sync.WaitGroup) {
	defer wg.Done()

	for {
		m.mutex.Lock()
		for !m.disponivel || m.ingredientes[meuIngrediente] {
			if m.rodadasAcabaram {
				m.mutex.Unlock()
				return
			}
			m.cond.Wait()
		}

		if !m.ingredientes[meuIngrediente] {
			fmt.Printf("Fumante %d (que tem %s) pegou %s e %s e está fumando.\n",
				id, ingredientNames[meuIngrediente], ingredientNames[(meuIngrediente+1)%3], ingredientNames[(meuIngrediente+2)%3])

			m.ingredientes[(meuIngrediente+1)%3] = false
			m.ingredientes[(meuIngrediente+2)%3] = false
			m.disponivel = false
			m.cond.Broadcast()
			m.mutex.Unlock()

			time.Sleep(time.Duration(rand.Intn(3)+1) * time.Second)
			fmt.Printf("Fumante %d terminou de fumar.\n", id)
		} else {
			m.mutex.Unlock()
		}
	}
}

func main() {
	mesa := NewMesa()
	var wg sync.WaitGroup

	wg.Add(4)

	go func() {
		defer wg.Done()
		mesa.colocarIngredientes()
	}()

	for i, ing := range []Ingredient{Tabaco, Papel, Fosforo} {
		go func(id int, ing Ingredient) {
			mesa.fumante(id, ing, &wg)
		}(i+1, ing)
	}

	wg.Wait()
	fmt.Println("Simulação concluída após 5 rodadas.")
}
